### PR TITLE
feat: encode struct-like arguments as tuples for ape-ethereum

### PR DESCRIPTION
**Pull request by fubuloubu**
### What I did

Allows any struct-like object (either dict or accessor via getattr) to be used whenever a struct input is required, instead of requiring a conversion to a tuple

fixes: #661

fixes: APE-464
### How I did it

I put it into ape-ethereum only because this is a bit more specific to how Ethereum's ABI encoding works. Wasn't sure if Starknet would allow the same

### How to verify it

Use it in a project with eip712 structs passed as an arg maybe

### Checklist

- [x] All changes are completed
- [x] New test cases have been added
- [ ] Documentation has been updated

### CODE REVIEW

File: tests/functional/conversion/test_encode_structs.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
+from ethpm_types.abi import MethodABI
+from hexbytes import HexBytes
+from pydantic import BaseModel
+
+from ape import networks
+
+ABI = MethodABI.parse_obj(
```
Comment (unparalleled-js) [inline tests/functional/conversion/test_encode_structs.py:7]: Does this align with both vyper and solidity ABI outputs?

File: tests/functional/conversion/test_encode_structs.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    "0000000000000000000000000000000000000000000000000000000000000001"
+    "3200000000000000000000000000000000000000000000000000000000000000"
+    "0000000000000000000000000000000000000000000000000000000000000001"
+)
+
+
+def test_encode_structs():
+    encode_calldata = networks.ethereum.encode_calldata
+
+    assert EXPECTED == encode_calldata(ABI, (1, b"2", True))
```
Comment (unparalleled-js) [inline tests/functional/conversion/test_encode_structs.py:42]: the "actual" goes first I believe, for some pytest output to make sense.

```suggestion
    assert encode_calldata(ABI, (1, b"2", True)) == EXPECTED
```

File: tests/functional/conversion/test_encode_structs.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    assert EXPECTED == encode_calldata(ABI, {"a": 1, "b": b"2", "c": True})
```
Comment (unparalleled-js) [inline tests/functional/conversion/test_encode_structs.py:43]:
```suggestion
    assert encode_calldata(ABI, {"a": 1, "b": b"2", "c": True}) == EXPECTED
```

File: tests/functional/conversion/test_encode_structs.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    assert EXPECTED == encode_calldata(ABI, Struct(a=1, b=b"2", c=True))
```
Comment (unparalleled-js) [inline tests/functional/conversion/test_encode_structs.py:44]:
```suggestion
    assert encode_calldata(ABI, Struct(a=1, b=b"2", c=True)) == EXPECTED
```

File: tests/functional/conversion/test_encode_structs.py [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+                    {"name": "b", "type": "bytes32"},
+                    {"name": "c", "type": "bool"},
+                ],
+            }
+        ],
+    }
+)
+
+
+class Struct(BaseModel):
```
Comment (unparalleled-js) [inline tests/functional/conversion/test_encode_structs.py:26]: Ethereum does not use `https://github.com/ApeWorX/ethpm-types/blob/main/ethpm_types/abi.py#L394` does it?

Comment (fubuloubu) [reply tests/functional/conversion/test_encode_structs.py:26]: No, not yet

**Review by unparalleled-js — APPROVED** [review]

**Review by unparalleled-js — CHANGES_REQUESTED** [review]
One thing is missing!

During testing, I noticed the conversion is failing. I am unable to use `AccountAPI` as an input in a dictionary for an input.

File: tests/functional/conversion/test_encode_structs.py [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+    a: int
```
Comment (unparalleled-js) [inline tests/functional/conversion/test_encode_structs.py:27]: can we make one of these an `AddressType` and include a test that passes a test account an input to ensure conversion works?

File: src/ape_ethereum/ecosystem.py [hunk 7]
```diff
@@ -320,6 +320,23 @@ def encode_calldata(self, abi: Union[ConstructorABI, MethodABI], *args) -> HexBy
...
+                and not isinstance(arg, tuple)
+            ):
+                if isinstance(arg, dict):
+                    args[idx] = tuple([arg[m.name] for m in ipt.components])  # type: ignore[index]
+
+                else:
+                    arg = [getattr(arg, m.name) for m in ipt.components]  # type: ignore[arg-type]
+                    args[idx] = tuple(arg)  # type: ignore[index]
+
         input_types = [i.canonical_type for i in abi.inputs]
```
Comment (unparalleled-js) [inline src/ape_ethereum/ecosystem.py:340]: do we also need to handle nested / recursive structures?

Comment (fubuloubu) [reply src/ape_ethereum/ecosystem.py:357]: I think `.canonical_type` is recursive

Comment (fubuloubu) [reply src/ape_ethereum/ecosystem.py:357]: https://github.com/ApeWorX/ethpm-types/blob/main/ethpm_types/abi.py#L58

Comment (unparalleled-js) [reply src/ape_ethereum/ecosystem.py:357]: I moreso meant struture values as inputs, like nested dicts with account objects in the middle:

```
account = accounts[0]
argument = {"foo": {"bar": account}}
contract.my_method(arguments, sender=account)
```

File: src/ape_ethereum/ecosystem.py [hunk 8]
```diff
@@ -337,8 +338,39 @@ def encode_calldata(self, abi: Union[ConstructorABI, MethodABI], *args) -> HexBy
...
+                    arguments[idx] = tuple([arg[m.name] for m in ipt.components])
+
+                else:
+                    arg = [getattr(arg, m.name) for m in ipt.components if m.name]
+                    arguments[idx] = tuple(arg)
+
         input_types = [i.canonical_type for i in abi.inputs]
-        encoded_calldata = encode(input_types, args)
+
+        def convert(ls):
```
Comment (unparalleled-js) [inline src/ape_ethereum/ecosystem.py:360]: idk, is there a better way to do this @fubuloubu ?
I tried using the tuple converter but it attempts to convert too much.

**Review by unparalleled-js — APPROVED** [review]
Approving, but waiting until at least @fubuloubu looks at the changes I added

**Review by fubuloubu — COMMENTED** [review]
Looking good!

**Pull request merged**